dts: bindings: Add lots more documentation

ulfalizer opened the pull request:
Two commits:

```
dts: bindings: Add lots more documentation

Add some information that would've saved me a lot of time:

 - Give an overview of how device tree nodes and bindings fit together,
   with examples. Assume people might be coming at it without knowing
   anything about device tree.

 - Explain how 'inherits' works in more detail

 - Explain what 'parent/child: bus: ...' does more concretely, and what
   problem it solves

 - Add more examples to show what things look like in the .dts file

 - Clean up the language a bit and make things more consistent

Also fix some errors, like 'properties: compatible: ...' being wrong
(missing 'constraint:' and compatible strings in the wrong place).
```

```
dts: bindings: Rename binding template to binding-template.yaml

More guessable than device_node.yaml.template.
```

ulfalizer commented:
This is based on https://github.com/zephyrproject-rtos/zephyr/pull/17740, so it should go in first.

carlescufi commented on dts/bindings/device_node.yaml.template:
```diff
@@ -1,88 +1,149 @@
-title: <short description of the node>
+# Overview of concepts (template below)
```
Could we include this documentation in `.rst` format as part of the compiled doc? I think it would probably be better than having it here in  the template

ulfalizer replied:
Thought the same, but got lazy while trying to figure out where to put it.

Could look at it tomorrow.

ulfalizer replied:
Moved the overview to the device tree page.

ulfalizer replied:
@carlescufi ping

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -71,11 +71,15 @@ The device tree files are compiled using the device tree compiler.  The compiler
 information is placed in a header file that is used by the rest of the code as
 the project is compiled.
 
+See the file ``dts/bindings/device_node.yaml.template`` for a binding template.
```
You can create a clickable link to the file by writing this as:
```
:zephyr_file:`dts/bindings/device_node.yaml.template` 
```

ulfalizer replied:
Fixed. I added links to `dts/bindings/` and `scripts/dts/` too.

ulfalizer replied:
@dbkinder ping

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -374,8 +378,8 @@ value or the format of its name.  Node properties are generated as C-style
 
-A YAML template file is provided to show the required format.  This file is
-located at::
+The binding format is documented in this file, which also has detailed
```
Agreed that the template should  be kept cleaner and most of the detail documentation should be put in a new document here, maybe ``dts_binding_format.rst``, or as an include file (as was done for flash partitions).

ulfalizer replied:
I integrated it into the existing documentation.

The template is really the meat of the documentation in the sense that you have to read it to make sense of bindings, so it felt a bit weird to hide it behind a link. It's now included directly in the generated documentation.

ulfalizer replied:
@dbkinder ping

ulfalizer committed:
dts: bindings: Rename binding template to binding-template.yaml

More guessable than device_node.yaml.template.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

erwango commented on dts/bindings/device_node.yaml.template:
```diff
@@ -1,88 +1,103 @@
+child:
+    bus: <string describing bus type, e.g. "i2c">
 
+# Used to simplify cases where a node has children that can all use the same
```
"Sub-nodes are used to ..." would make it clearer.

ulfalizer replied:
Fixed

erwango reviewed (changes requested):
Nitpick, otherwise it's a welcome addition

ulfalizer committed:
dts: bindings: Add lots more documentation

Add some information that would've saved me a lot of time:

 - Give an overview of how device tree nodes and bindings fit together,
   with examples. Assume people might be coming at it without knowing
   anything about device tree.

 - Explain how 'inherits' works in more detail

 - Explain what 'parent/child: bus: ...' does more concretely, and what
   problem it solves

 - Add more examples to show what things look like in the .dts file

 - Clean up the language a bit and make things more consistent

Also fix some errors, like 'properties: compatible: ...' being wrong
(missing 'constraint:' and compatible strings in the wrong place).

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
Now rebased to not require https://github.com/zephyrproject-rtos/zephyr/pull/17740. Should be fine to merge after review.

alexanderwachter commented on doc/guides/dts/index.rst:
```diff
@@ -356,32 +359,57 @@ Example: Subset of DTS/YAML files for NXP FRDM K64F (Subject to Change)::
-data from the device tree nodes.  For this, we rely on YAML binding files to
-describe the contents or definition of a device tree node and instruct how the
-extracted information should be formatted.
+Binding are stored in :zephyr_file:`dts/bindings/`.
```
Maybe a word that the filename matches the compatible string?

ulfalizer replied:
Added (and s/binding/bindings/)

carlescufi reviewed (approved):
Thanks for the updates, this LGTM now!

dbkinder reviewed (approved):
doc changes look OK, thanks!
